internal/providers/aws/aws.go

```diff
@@ -73,3 +76,43 @@ func Init(f *resource.Fetcher) error {
 	f.S3RegionHint = regionHint
 	return nil
 }
+
+// fetchFromAWSMetadata fetches metadata from the `IMDSv2` service if its
+// configured, else it will fall back to `IMDSv1`.
+func fetchFromAWSMetadata(u url.URL, opts resource.FetchOptions, f *resource.Fetcher) ([]byte, error) {
+	token, err := fetchAWSIMDSV2Token(f)
+	if err == errIMDSV2 {
+		// Do nothing
+		f.Logger.Info("IMDSv2 service is unavailable; falling back to IMDSv1")
+	} else if err != nil {
+		return nil, err
+	} else {
+		opts.Headers.Add("X-aws-ec2-metadata-token", token)
+	}
+	return f.FetchToBuffer(u, opts)
+}
+
+// fetchAWSIMDSV2Token fetches a session token from an EC2 instance (if the
+// instace is configured to use `IMDSv2`), otherwise, it will return an error.
+func fetchAWSIMDSV2Token(f *resource.Fetcher) (string, error) {
+	imdsTokenURL := url.URL{
```

Comment:
Hmm, how did this end up down here?  Please move it back to the `var` block.

Reply:
I think we're using this var only in `fetchAWSIMDSV2Token`, so I'm a little hesitant to include that as a module var.

Reply:
Well, the same is true of `userdataURL`.  The idea is to keep significant constants in one place at the top of the file.

Reply:
ok